README.md
# trading_returns_analysis

## Table of Contents

- [About](#about)
- [Getting Started](#getting_started)
- [Usage](#usage)
- [Contributing](../CONTRIBUTING.md)

## About <a name = "about"></a>

trading_returns _analysis modules contains function to analyze a timeseries of trading returns. Such as, cumulative returns monthly returns, etc.


## Getting Started <a name = "getting_started"></a>

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See [deployment](#deployment) for notes on how to deploy the project on a live system.

### Prerequisites

First, you need to create a conda virtual environment together with python version 3.9.5 and at the same time install the dependencies in the requirements.txt file.

### Windows CMD Terminal
```
conda create --name TypeYourVirtualEnvironmentHere python=3.9.5 --file requirements.txt

```
Next, activate the virtual environment that you just created now. In the windows terminal, type the following commands.

### Windows CMD Terminal
```
conda activate TypeYourVirtualEnvironmentHere

```
### Installing

Next, after you have created a conda virtual environment with python version 3.9.5 together with the dependencies in the requirements.txt, you need to pip install sqlconnection (the "Module"). In the windows terminal, type the following codes below.

### Windows CMD Terminal
```
pip install version pip install git+https://github.com/Iankfc/trading_returns_analysis.git@master --no-deps
```

To use the module in a pythone terminal, import the module just like other python modules such as pandas or numpy.

### Python Terminal
```

from asset_price_etl.Download_Scripts import etl_fx_histadata_001 as etl
import trading_exit_price as tep
from trading_direction import func_list_str_generate_random_trades as td


df_data = etl._function_extract(_str_valuedate_start = '1/1/2011',
                                    _str_valuedate_end = '12/31/2011',
                                    _str_resample_frequency = 'D',
                                    str_currency_pair = 'EURUSD')

#%% Create a simulated trade based on random number generation between 0 & 2
df_data['TradeDirection'] = td.func_list_str_generate_random_trades(df_data)

df_data['TakeProfitRate'] = 0.01
df_data['StoplossRate'] = 0.005

df_data['TimeFrame'] = pd.NA
df_data['Broker'] = pd.NA
df_data['StrategyName'] = pd.NA
df_data['OtherParametersJSON'] = np.array({},dtype = object)
df_data['InsertDateTime'] = pd.NaT
df_data['PythonFilePath'] = pd.NA


df_data = func_df_plotlychart_generate_returns_analysis(df_data = df_data,
                                                        str_column_trade_entry_price_column_name = 'Open',
                                                        str_column_trade_direction_column_name = 'TradeDirection',
                                                        int_initial_balance_in_usd = 10_000,
                                                        float_percent_risk_per_trade = 0.01,
                                                        bool_appy_kelly_criterion_True_or_False = True,
                                                        float_kelly_criterion_multiplier = 0.1,
                                                        str_stoploss_fix_or_variable = 'variable',
                                                        bool_apply_CumulativeReturnReverseIndicatorSMA_True_or_False = True,
                                                        str_rolling_return_sampling_duration_for_trade_hold_or_reverse = '30D',
                                                        bool_reverse_trade_direction_when_cumulative_return_trend_down_True_or_False = False,
                                                        bool_hold_trade_when_cumulative_return_trend_down_True_or_False = True,
                                                        bool_interchange_sl_and_tp_when_cumulative_return_trend_down_True_or_False = True)
            

```




CHANGELOG.md

# Change Log
All notable changes to this project will be documented in this file.

## v3.2.1 - 2021-12-30
  Bug fix - Change OtherParametersJSON to TradingParametersJSON column
## v3.2.0 - 2021-12-15
  Add feature - new function to generate a performance summary of all trading simulations
## v3.1.0 - 2021-12-13
  Add feature - Upload performance into SQL Server database
## v3.0.0 - 2021-12-10
  Bug fix - Incorrect cumulative return which includes future unrealized return. 

## v2.0.2 - 2021-12-07
  Big fix - Error in creating rolling return with data less than the rolling duration
  
## v2.0.1 - 2021-12-07
  Fix Bug - Incorrect function argument for exit price

## v2.0.0 - 2021-11-24
  Maintenance - Update python dependencies

## v1.6.1 - 2021-09-21
  Fix bug - Missing option to reverse or hold lossing trades
## v1.6.0 - 2021-09-21
  Add feature - Incorporate trading_exit_price module inside the trading_returns_analysis module

## v1.5.2 - 2021-09-20
  Fix bug - missing trading_exit_price module
## v1.5.1 - 2021-09-20
  Fix bug - fix plotly chart variable
## v1.5.0 - 2021-09-19
  Add feature - Combine data frame and plotly chart

## v1.4.0 - 2021-09-15
  Add feature - add time lag to consider/disregard unrealised return
## v1.3.0 - 2021-09-15
  Add feature to hold or reverse trade based on rolling returns
## v1.2.0 - 2021-09-13

  Fix bug Incorrect cumulative rate
## v1.2.0 - 2021-09-13

  Add feature - Apply kelly criterion position in usd adjustment
## v1.1.1 - 2021-09-09

  Add feature - Addition of cumulative balance given an initial balance and risk per trade
## v1.0.1 - 2021-09-09

  Fix bug - Correct versioni of dependencies, trading_exit_price
  

## v1.0.0 - 2021-09-09
 
  Intial release



trading_returns_analysis/trading_simulation_summary.py
#%% Import modules

import sqlserverconnection as sc
import os
import pandas as pd

#%%

def generate_trading_simulations_summary():
    str_sqlquery_file = 'trading_simulation_summary.sql'
    str_sql_query = open(fr"{os.path.dirname(__file__)}/{str_sqlquery_file}",'r').read()

    obj_sql_connection  = sc.CONNECT_TO_SQL_SERVER(_str_server = "localhost",
                                                _str_database = 'db_oms',
                                                _str_trusted_connection = 'no',
                                                str_download_or_upload = 'download')

    df_data = pd.read_sql(sql = str_sql_query,
                    con = obj_sql_connection
                    )


    return df_data

#%%
if __name__ == '__main__':
    df_data = generate_trading_simulations_summary()

#%%


trading_returns_analysis/trading_simulation_template1.py
# -*- coding: utf-8 -*-
"""
Created on Sun Feb  7 10:36:05 2021

@author: recon
"""

#%% Set Variables

_str_currency = 'EURUSD'
_str_timeframe = 'D'
_int_hours_offset = 3
_float_takeprofit_rate = 0.0100
_float_stoploss_rate = 0.0030

#%% Import modules


from numba.core.errors import HighlightColorScheme
import pandas as pd
import numpy as np
from numba import njit
import warnings
from sqlconnection import CONNECT_TO_SQL_SERVER as _module_sc
from asset_price_etl import etl_fx_histadata_001 as etl
import __init__ as tradingformula
import statistics
from scipy import stats
import plotly.graph_objects as go
import plotly.io as pio
from plotly.subplots import make_subplots



#%% Remove warnings
warnings.filterwarnings('ignore')


#%% ETL load fx eurusd historical price

df_fx_raw_data = etl._function_extract(_str_valuedate_start = '1/1/2010',
                                        _str_valuedate_end = '12/31/2010')


#%% Sort datetimie into ascending order and set datetime as the index

df_fx_raw_data = df_fx_raw_data.sort_values('DateTime', ascending = True)

df_fx_raw_data = df_fx_raw_data.set_index('DateTime')

#%% Resample price into desire timeframe

df_fx_raw_data = df_fx_raw_data.resample(_str_timeframe).agg({'Open':'first',
                                                                'High':'max',
                                                                'Low':'min',
                                                                'Close':'last'
                                                                })


#%% Create a copy of the raw data 

df_fx = df_fx_raw_data.copy()

#%% Calculate the daily percentage from high to low and take the absolute value
df_fx['VolatilityRank'] = tradingformula.GENERATE_HIGH_TO_LOW_VOLATILITY_SCORE(_df_data = df_fx,
                                                                                _variant_number_of_period = '2D',
                                                                                _str_expanding_or_rolling_historical_volatility = 'rolling',
                                                                                _int_rolling_number_of_days = 30,
                                                                                _str_column_high = 'High',
                                                                                _str_column_low = 'Low')

#%% Create a plotly plot that shows the fx price together with the volatility

_obj_fig = make_subplots(rows = 2, cols=1, shared_xaxes=True)


_obj_fig.add_trace(go.Candlestick(
                                    x = df_fx.index,
                                    open = df_fx.Open,
                                    high = df_fx.High,
                                    low = df_fx.Low,
                                    close = df_fx.Close
                                    ),
                                row = 1,
                                col=1
                                )

_obj_fig.add_trace(go.Scatter(x = df_fx.index,y = df_fx.VolatilityRank),
                    row = 2,
                    col = 1)

_obj_fig.update_layout(xaxis_rangeslider_visible=False)

pio.renderers.default = 'browser'

pio.show(_obj_fig)

#%% Create a random sample of Long Short Hold

df_fx['TradeDirection'] = tradingformula.GENERATE_RANDOM_TRADE_DIRECTION(_df_data = df_fx)


#%% This is an example of volatility filter on when to trade and when not to trade at a specific volatility rank

df_fx['TradeDirection'] = np.where((df_fx['VolatilityRank'] >= 20) & (df_fx['VolatilityRank'] <= 80), 
                                   df_fx['TradeDirection'],
                                   'Hold')



#%% Create a Long take profit of 2%

df_fx['TakeProfitPrice'] = tradingformula.GENERATE_FIX_TAKE_PROFIT_PRICE(_df_data = df_fx,
                                                                         _float_takeprofit_rate = _float_takeprofit_rate)

#%% Add column for Stoploss Price Relative to the open price

df_fx = tradingformula.GENERATE_FIX_AND_SEPARATE_LONG_SHORT_STOPLOSS(_df_data = df_fx,
                                                                     _str_column_Open_price = 'Open',
                                                                     _float_stoploss_rate = _float_stoploss_rate)

#%% Create a new column that determines if the previous i
#period created a higher high or lower high or higher low or lower low

df_fx = tradingformula.GENERATE_HIGHER_HIGH_LOWER_LOW_COLUMN(data = df_fx,
                                                             _column_high_price = 'High',
                                                             _column_row_price = 'Low')
            

#%%

df_fx = tradingformula.GET_EXIT_PRICE_BASED_ON_TRAILING_STOP(data = df_fx,
                                                            _column_high_price = 'High',
                                                            _column_low_price = 'Low',
                                                            _column_TradeDirection = 'TradeDirection',
                                                            _column_LongStopLossRelativetoOpenPrice = 'LongStopLossRelativetoOpenPrice',
                                                            _column_ShortStopLossRelativetoOpenPrice = 'ShortStopLossRelativetoOpenPrice')
    
#%% Identify Static Take Profit Hit Date

df_fx = tradingformula.GET_TAKE_PROFIT_AND_DATE(data = df_fx,
                                                 _column_high_price = 'High',
                                                 _column_low_price = 'Low',
                                                 _column_TakeProfitPrice = 'TakeProfitPrice',
                                                 _column_TradeDirection = 'TradeDirection')

#%% Generate an exit price based on which was triggered first, is it the stoploss or the take profit

df_fx = tradingformula.GENERATE_FINAL_EXIT_PRICE(data = df_fx,
                                                 _column_TakeProfitHitDate = 'TakeProfitHitDate',
                                                 _column_TakeProfitPrice = 'TakeProfitPrice',
                                                 _column_TrailingExitStopLossDate = 'TrailingExitStopLossDate',
                                                 _column_TrailingStoplossExitPrice = 'TrailingStoplossExitPrice')

#%% Returns analysis gives you the cumulative return so as the rolling return

df_fx = tradingformula.RETURNS_ANALYSIS(data = df_fx,
                                          _column_trade_entry_price ='Open',
                                          _column_trade_direction = 'TradeDirection',
                                          _column_trade_exit_price = 'ExitPrice')
    


#%% Create new column that will determine the cumulative win rate and the rolling win rate


df_fx = tradingformula.CUMULATIVE_AND_ROLLING_WIN_RATE(df_fx, 'SingleTradePercentageChange')


#%% Create new column that will determine the cumulative risk and return and the rolling win rate


df_fx = tradingformula.CUMULATIVE_AND_ROLLING_RISK_TO_RETURN_RATIO(df_fx, 'SingleTradePercentageChange')


#%% Creates a new column that gives you the rate at which you need to increase or decrease your postion size

df_fx = tradingformula.CUMULATIVE_KELLY_CRITERION(df_fx, 'SingleTradePercentageChange')




#%% Generate a csv output

df_fx.to_csv('OutputManualSenseCheck.csv')



    
    
    
    
    
    
    
    
    